domain/application/state/exposed.go

```diff
@@ -5,38 +5,481 @@ package state
+			return errors.Errorf("preparing unset exposed cidr endpoint %q on application %q query: %w", endpoint, appID, err)
+		}
+		if err := tx.Query(ctx, unsetExposedCIDRStmt, applicationID, endpointName).Run(); err != nil {
+			return errors.Errorf("unsetting exposed cidr endpoint %q on application %q: %w", endpoint, appID, err)
+		}
+	}
+
+	// Since we need to keep referential integrity with respect to the endpoint
+	// as stored in charm_relation, we first check if the provided endpoint is
+	// the wildcard and in that case we simply remove the spaces where the
+	// application_endpoint_uuid is NULL.
+	var (
+		unsetExposedSpaceQuery string
+		unsetExposedSpaceStmt  *sqlair.Statement
+	)
+	if endpoint == wildcardEndpointName {
+		unsetExposedSpaceQuery = `
+DELETE FROM application_exposed_endpoint_space
+WHERE application_uuid = $applicationID.uuid
+AND application_endpoint_uuid IS NULL;
+`
+		unsetExposedSpaceStmt, err = st.Prepare(unsetExposedSpaceQuery, applicationID)
+		if err != nil {
+			return errors.Errorf("preparing unset exposed space endpoint %q on application %q query: %w", endpoint, appID, err)
+		}
+		if err := tx.Query(ctx, unsetExposedSpaceStmt, applicationID).Run(); err != nil {
+			return errors.Errorf("unsetting exposed space endpoint %q on application %q: %w", endpoint, appID, err)
+		}
+	} else {
+		unsetExposedSpaceQuery = `
+DELETE FROM application_exposed_endpoint_space
+WHERE application_uuid = $applicationID.uuid 
+AND application_endpoint_uuid IN (
+    SELECT application_endpoint_uuid
+    FROM application_endpoint
+    JOIN charm_relation 
+        ON application_endpoint.charm_relation_uuid = charm_relation.uuid
+    WHERE charm_relation.name = $setEndpointName.name
+    AND application_endpoint.application_uuid = $applicationID.uuid
+);
+`
+		unsetExposedSpaceStmt, err = st.Prepare(unsetExposedSpaceQuery, applicationID, endpointName)
+		if err != nil {
+			return errors.Errorf("preparing unset exposed space endpoint %q on application %q query: %w", endpoint, appID, err)
+		}
+		if err := tx.Query(ctx, unsetExposedSpaceStmt, applicationID, endpointName).Run(); err != nil {
+			return errors.Errorf("unsetting exposed space endpoint %q on application %q: %w", endpoint, appID, err)
+		}
+	}
+
+	return nil
+}
+
+func (st *State) upsertExposedSpaces(ctx context.Context, tx *sqlair.TX, appID coreapplication.ID, endpoint string, exposeToSpaceIDs set.Strings) error {
+	if exposeToSpaceIDs.Size() == 0 {
+		return nil
+	}
+
+	var upsertExposedSpaceQuery string
+
+	// Since we need to keep referential integrity with respect to the endpoint
+	// as stored in charm_relation, we first check if the provided endpoint is
+	// the wildcard and in that case we simply insert the spaces and let the
+	// endpoint NULL.
+	if endpoint == wildcardEndpointName {
+		upsertExposedSpaceQuery = `
+INSERT INTO application_exposed_endpoint_space(application_uuid, space_uuid)
+VALUES ($setExposedSpace.application_uuid, $setExposedSpace.space_uuid)
```

Comment:
Do you not care about conflicts?

Reply:
No, we don't care because what I've done is to remove all the spaces (or cidrs) that for a given application_uuid and application_endpoint_uuid so there cannot be any conflicts.
That being said, I do realize now that I can get rid of the `ON CONFLICT` clause that I have in the query to insert non-wildcard endpoints.

Reply:
If we end up watching this, then we'll get a double pump on the watcher. One for remove, one for add.

Keep this in mind.

Reply:
Yeah good call, but I don't see how I can ensure a correct merge without previously removing all rows for a given app and endpoint. If you have a better idea I'm happy to change it.

Reply:
You have to build a set of additions, updates and deletions. Then only change them based on that set.

Reply:
Discussed offline: this is a non-issue since the changes under the same transaction will always end up in the same term (for a notify watcher at least) and due to coalescing, we can be sure that multiple deletes and inserts under the same tx will only generate one notify.